src/Nethermind/Nethermind.Init/Steps/EthereumStepsManager.cs

```diff
@@ -36,153 +34,102 @@ public EthereumStepsManager(
         {
-            while (_allSteps.Any(static s => s.Stage != StepInitializationStage.Complete))
+            List<Task> allRequiredSteps = CreateAndExecuteSteps(cancellationToken);
+            if (allRequiredSteps.Count == 0)
+                return;
+            do
             {
-                cancellationToken.ThrowIfCancellationRequested();
-
-                RunOneRoundOfInitialization(cancellationToken);
-                await ReviewDependencies(cancellationToken);
-                ReviewFailedAndThrow();
-            }
-
-            await Task.WhenAll(_allPending);
+                Task current = await Task.WhenAny(allRequiredSteps);
+                ReviewFailedAndThrow(current);
+                if (current.IsCanceled && _logger.IsDebug)
+                    _logger.Debug($"A required step was cancelled!");
+                allRequiredSteps.Remove(current);
+            } while (allRequiredSteps.Any(s => !s.IsCompleted));
         }
 
-        private readonly ConcurrentQueue<Task> _allPending = new();
 
-        private void RunOneRoundOfInitialization(CancellationToken cancellationToken)
+        private List<Task> CreateAndExecuteSteps(CancellationToken cancellationToken)
         {
-            int startedThisRound = 0;
+            Dictionary<Type, List<StepWrapper>> stepBaseTypeMap = [];
+            Dictionary<Type, StepInfo> stepInfoMap = [];
+
             foreach (StepInfo stepInfo in _allSteps)
             {
                 cancellationToken.ThrowIfCancellationRequested();
 
-                if (stepInfo.Stage != StepInitializationStage.WaitingForExecution)
-                {
-                    continue;
-                }
-
                 IStep? step = CreateStepInstance(stepInfo);
                 if (step is null)
+                    throw new StepDependencyException($"A step {stepInfo} could not be created and initialization cannot proceed.");
+                stepInfoMap.Add(step.GetType(), stepInfo);
+                if (stepBaseTypeMap.ContainsKey(stepInfo.StepBaseType))
                 {
-                    if (_logger.IsError) _logger.Error($"Unable to create instance of Ethereum runner step {stepInfo}");
-                    continue;
-                }
-
-                if (_logger.IsDebug) _logger.Debug($"Executing step: {stepInfo}");
-
-                stepInfo.Stage = StepInitializationStage.Executing;
-                startedThisRound++;
-                Task task = ExecuteStep(step, stepInfo, cancellationToken);
-
-                if (step.MustInitialize)
-                {
-                    _allPending.Enqueue(task);
+                    stepBaseTypeMap[stepInfo.StepBaseType].Add(new StepWrapper(step));
                 }
                 else
                 {
-                    stepInfo.Stage = StepInitializationStage.Complete;
+                    stepBaseTypeMap.Add(stepInfo.StepBaseType, new List<StepWrapper>() { new StepWrapper(step) });
```

Comment:
You can do it sligthly better with [CollectionsMarshal.GetValueRefOrNullRef](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.collectionsmarshal.getvaluerefornullref?view=net-9.0#system-runtime-interopservices-collectionsmarshal-getvaluerefornullref-3(system-collections-generic-dictionary((-0-1))-alternatelookup((-2))-2))

```
List<StepWrapper>? list = CollectionsMarshal.GetValueRefOrNullRef(stepBaseTypeMap, stepInfo.StepBaseType);
list ??= new();
list.Add(new StepWrapper(step))
```